src/audio/base_fw.c

```diff
@@ -419,6 +421,27 @@ static int basefw_power_state_info_get(uint32_t *data_offset, char *data)
 	return 0;
 }
 
+static int basefw_modules_info_get(uint32_t * data_offset, char* data)
+{
+	struct ipc4_modules_info *const module_info = (struct ipc4_modules_info *)data;
+
+	struct sof_man_fw_desc *desc = (struct sof_man_fw_desc *)IMR_BOOT_LDR_MANIFEST_BASE;
+	if (!desc) {
+		return -EINVAL;
+	}
+
+	module_info->modules_count = desc->header.num_module_entries;
+
+	for(int idx = 0; idx < module_info->modules_count;  ++idx){
+		struct sof_man_module *module_entry =
+			(struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(idx));
+		memcpy(&(module_info->modules[idx]), module_entry, sizeof(struct sof_man_module));
+		}
+
+	*data_offset = sizeof(uint32_t) + module_info->modules_count * sizeof(struct sof_man_module);
```

Comment:
I'd do `sizeof(module_info->modules_count) + module_info->modules_count * sizeof(module_info->modules[0])` - makes it clearer what is calculated IMHO